### PR TITLE
(82)(68) validation and testing + Display friends

### PR DESCRIPTION
**Pull request by based-kwl**
- Validation class
- Added many tests for current API routes (67 tests total)
- Merged PR #73

### CODE REVIEW

File: __test__/student.spec.js [hunk 1]
```diff
@@ -19,133 +20,395 @@ afterAll(async () => {
...
+      .then((res) => {
+        assert.deepEqual(res.body.email, user2.email)
+        assert.deepEqual(res.body.friends, [user1.email, user3.email])
       })
   });
 
-  //works as cleanup of previous test too.
-  it("delete a Student ", async () => {
+  //Delete student tests
+  it("Delete a Student ", async () => {
```
Comment (mnashat) [inline __test__/student.spec.js:387]: After deleting a student (user1.email in this case), we should test that this student is also removed from the friend list of other students (user2) in this case.

File: __test__/student.spec.js [hunk 2]
```diff
@@ -19,133 +20,395 @@ afterAll(async () => {
...
+      .expect(401)
+      .then((res) => {
+        assert.deepEqual(res.body, { auth: false, message: 'No token found in request' })
+      })
+  });
 
-    let expected = "user_" + random4
-    let arr = ["user_" + random2 + "@gmail.com", "user_" + random3 + "@gmail.com", "user_" + random0 + "@gmail.com"]
+  //Friend request tests
+  let FriendRequest, FriendRequest2;
```
Comment (mnashat) [inline __test__/student.spec.js:202]: Please change variable names to start with lowercase characters.

**Review by mnashat — CHANGES_REQUESTED** [review]
Great job on the tests, well done! Just a few comments/changes for consistency. Check the comments.

Also, we should rename `Student_test_data.js` to `student_test_data.js` for consistency with other file names.

Thank you!

File: __test__/student.spec.js (hunk 1 above)
Comment (based-kwl) [reply __test__/student.spec.js:387]: Thus functionality does not exist. It has to be implemented first to test it

Comment (mnashat) [reply __test__/student.spec.js:387]: Noted, thanks for letting me know.

**Review by mnashat — APPROVED** [review]
Great job on the tests! Approved.

**Comment by guaj** [comment]
your branch is adding code that I deleted (duplicates). You should probably rebase and remove code push in the following files :
- friendRequest repository
- friends.js

Also implemented e2e tests to check if your features is breaking functionnality with main. All tests are failing right now, so need to check these out before merge.

**Review by guaj — CHANGES_REQUESTED** [review]

File: src/repository/friendRequestRepository.js [hunk 3]
```diff
@@ -2,14 +2,40 @@ const FriendRequest = require('../models/friendRequest.model');
 
 module.exports = class FriendRequestRepository {
 
+    /**
+     * Delete a friend request.
+     * @param {string} requestID : id of the request to delete
+     * @returns {Promise<unknown>}
+     */
+    static async deleteFriendRequest(requestID) {
```
[the remark this answers is not in the document]
Comment (based-kwl) [reply src/repository/friendRequestRepository.js:10]: This is not duplicate.findOneAndDelete is not the same as deleteOne. findOneAndDelete will return the document deleted while deleteOne will only return the status, i.e. number of documents deleted (1)

[the remark this answers is not in the document]
Comment (based-kwl) [reply src/repository/friendRequestRepository.js:10]: delete uses deleteOne and deleteFriendRequest uses findOneAndDelete

**Comment by based-kwl** [comment]
> your branch is adding code that I deleted (duplicates). You should probably rebase and remove code push in the following files :
> 
> * friendRequest repository
> * friends.js
> 
> Also implemented e2e tests to check if your features is breaking functionnality with main. All tests are failing right now, so need to check these out before merge. In the readMe file in the FE repo you'll find the procedure on how to run these.
> 
> image

resolved, see FE PR for more details

**Comment by based-kwl** [comment]
image
Results of e2e tests with https://github.com/based-kwl/SemesterPlanner-client/pull/94

**Pull request merged**